test/common/upstream/ring_hash_lb_test.cc

```diff
@@ -327,5 +327,89 @@ TEST_P(RingHashLoadBalancerTest, UnevenHosts) {
+  std::unordered_map<uint64_t, uint32_t> expected{
+      {928266305478181108UL, 2},  {4443673547860492590UL, 2},  {5583722120771150861UL, 1},
+      {6311230543546372928UL, 1}, {13444792449719432967UL, 2}, {16117243373044804889UL, 0}};
+
+  LoadBalancerPtr lb = lb_->factory()->create();
+  for (const auto& entry : expected) {
+    TestLoadBalancerContext context(entry.first);
+    EXPECT_EQ(hostSet().hosts_[entry.second], lb->chooseHost(&context));
+  }
+}
+
+TEST_P(RingHashLoadBalancerTest, LocalityWeighted) {
+  hostSet().hosts_ = {makeTestHost(info_, "tcp://127.0.0.1:90"),
+                      makeTestHost(info_, "tcp://127.0.0.1:91"),
+                      makeTestHost(info_, "tcp://127.0.0.1:92")};
+  hostSet().healthy_hosts_ = hostSet().hosts_;
+  hostSet().hosts_per_locality_ =
+      makeHostsPerLocality({{hostSet().hosts_[0]}, {hostSet().hosts_[1]}, {hostSet().hosts_[2]}});
+  hostSet().healthy_hosts_per_locality_ = hostSet().hosts_per_locality_;
+  // assign locality weights with a greatest common denominator greater than 1, to validate that the
+  // ring won't contain unnecessary duplicate entries.
+  hostSet().locality_weights_ = makeLocalityWeights({2, 4, 6});
+  hostSet().runCallbacks({}, {});
+
+  config_ = (envoy::api::v2::Cluster::RingHashLbConfig());
+  config_.value().mutable_minimum_ring_size()->set_value(6);
+  init();
+
+  // :90 should appear once, :91 should appear twice and :92 should appear three times.
+  std::unordered_map<uint64_t, uint32_t> expected{
+      {928266305478181108UL, 2},  {4443673547860492590UL, 2},  {5583722120771150861UL, 1},
+      {6311230543546372928UL, 1}, {13444792449719432967UL, 2}, {16117243373044804889UL, 0}};
+
+  LoadBalancerPtr lb = lb_->factory()->create();
+  for (const auto& entry : expected) {
+    TestLoadBalancerContext context(entry.first);
+    EXPECT_EQ(hostSet().hosts_[entry.second], lb->chooseHost(&context));
+  }
+}
+
+TEST_P(RingHashLoadBalancerTest, HostAndLocalityWeighted) {
+  hostSet().hosts_ = {makeTestHost(info_, "tcp://127.0.0.1:90", 1),
+                      makeTestHost(info_, "tcp://127.0.0.1:91", 2),
+                      makeTestHost(info_, "tcp://127.0.0.1:92", 3)};
+  hostSet().healthy_hosts_ = hostSet().hosts_;
+  hostSet().hosts_per_locality_ =
+      makeHostsPerLocality({{hostSet().hosts_[0]}, {hostSet().hosts_[1]}, {hostSet().hosts_[2]}});
+  hostSet().healthy_hosts_per_locality_ = hostSet().hosts_per_locality_;
+  hostSet().locality_weights_ = makeLocalityWeights({1, 2, 3});
+  hostSet().runCallbacks({}, {});
+
+  config_ = (envoy::api::v2::Cluster::RingHashLbConfig());
+  config_.value().mutable_minimum_ring_size()->set_value(0);
+  init();
+
+  // :90 should appear once, :91 should appear four times and :92 should appear nine times.
+  std::unordered_map<uint64_t, uint32_t> expected{
+      {928266305478181108UL, 2},   {4443673547860492590UL, 2},  {4470782202023056897UL, 1},
```

Comment:
What are these raw numbers? Is there something human readable we can use here instead?

Reply:
Yeah, tests that did more interesting disiributions would be great, I wouldn't scrap any existing tests while doing that, they probably are covering some behaviors that are important to folks today.

Reply:
Note: I still need to address this.

Reply:
I've added tests to show that with a high `target_hashes_per_host` configuration, the ring distribution sort of approximates the specified weighting.